pkg/ztunnel/iptables/inpod.go

```diff
@@ -0,0 +1,480 @@
+	InpodMask        = 0xfff
+	InpodRestoreMask = 0xffffffff
+
+	ZtunnelInboundPort          = 15008
+	ZtunnelOutboundPort         = 15001
+	ZtunnelInboundPlaintextPort = 15006
+
+	VersionSpecificPlaceholder = "<VERSION_SPECIFIC>"
+)
+
+// CreateInPodRules creates the iptables rules for ztunnels inpod mode.
+//
+// Note that this function is supposed to be called from within the pods
+// network namespace.
+func CreateInPodRules(logger *slog.Logger, ipv4Enabled, ipv6Enabled bool) error {
+	if err := addLoopbackRoute(logger, ipv6Enabled); err != nil {
+		return err
+	}
+
+	if err := addInPodMarkRule(ipv6Enabled); err != nil {
+		return err
+	}
+
+	if err := addInPodRules(logger, ipv4Enabled, ipv6Enabled); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addLoopbackRoute(logger *slog.Logger, ipv6Enabled bool) error {
+	// Set up netlink routes for localhost
+	cidrs := []string{"0.0.0.0/0"}
+	if ipv6Enabled {
+		cidrs = append(cidrs, "0::0/0")
+	}
+	for _, fullCIDR := range cidrs {
+		_, localhostDst, err := net.ParseCIDR(fullCIDR)
+		if err != nil {
+			return fmt.Errorf("parse CIDR: %w", err)
+		}
+
+		// Equiv: "ip route add local 0.0.0.0/0 dev lo table 100"
+		ciliumRoute := route.Route{
+			Device: "lo",
+			Prefix: *localhostDst,
+			Scope:  netlink.SCOPE_HOST,
+			Type:   unix.RTN_LOCAL,
+			Table:  RouteTableInbound,
+		}
+
+		if err := route.Upsert(logger, ciliumRoute); err != nil {
+			return fmt.Errorf("failed to add route (%+v): %w", ciliumRoute, err)
+		}
+	}
+	return nil
+}
+
+func addInPodMarkRule(ipv6Enabled bool) error {
+	mask := uint32(InpodMask)
+
+	// IPv4 rule
+	ipv4Rule := route.Rule{
+		Priority: InpodRulePriority,
+		Mark:     InpodTProxyMark,
+		Mask:     mask,
+		Table:    RouteTableInbound,
+	}
+
+	if err := route.ReplaceRule(ipv4Rule); err != nil {
+		return fmt.Errorf("failed to configure IPv4 netlink rule: %w", err)
+	}
+
+	if ipv6Enabled {
+		// IPv6 rule
+		ipv6Rule := route.Rule{
+			Priority: InpodRulePriority,
+			Mark:     InpodTProxyMark,
+			Mask:     mask,
+			Table:    RouteTableInbound,
+		}
+
+		if err := route.ReplaceRuleIPv6(ipv6Rule); err != nil {
+			return fmt.Errorf("failed to configure IPv6 netlink rule: %w", err)
+		}
+	}
+	return nil
+}
+
+type rule struct {
+	ipv4       string
+	ipv6       string
+	table      string
+	chain      string
+	parameters []string
+}
+
+type ruleManager struct {
+	rules  []rule
+	logger *slog.Logger
+	ipt4   *iptables.IPTables
+	ipt6   *iptables.IPTables
+}
+
+func (m *ruleManager) add(table, chain string, parameters ...string) {
+	m.rules = append(m.rules, rule{"", "", table, chain, parameters})
+}
+
+func (m *ruleManager) addVersioned(ipv4, ipv6, table, chain string, parameters ...string) {
+	m.rules = append(m.rules, rule{ipv4, ipv6, table, chain, parameters})
+}
+
+func replaceIPPlaceholder(args []string, ip string) []string {
+	if ip == "" {
+		return args
+	}
+	res := make([]string, len(args))
+	copy(res, args)
```

Comment:
nit, you can do this in one line with `slice.Clone` which really just does 
```go
return append(S{}, s...)
```
under the hood.